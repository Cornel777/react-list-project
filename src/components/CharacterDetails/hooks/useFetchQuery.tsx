import { useQuery, keepPreviousData } from "@tanstack/react-query";
import fetchSwapi from "../../../services/api/fetchSwapi";

export interface HookReturn<T> {
  isFetching: boolean;
  isPlaceholderData: boolean;
  isPending: boolean;
  isError: boolean;
  error: Error | null;
  data: T | undefined;
}


const useFetchQuery = <T, >(queryKey: string, path: string, id: string | undefined): HookReturn<T> => {

  const { isFetching, isPlaceholderData, isPending, isError, error, data } = useQuery<T>({
    queryKey: [queryKey, id],
    queryFn: () => fetchSwapi(path),
    staleTime: 5000,
    placeholderData: keepPreviousData,
  })

    return {
      isFetching,
      isPlaceholderData,
      isPending,
      isError,
      error,
      data,
    }
}

export default useFetchQuery;