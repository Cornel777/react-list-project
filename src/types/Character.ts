export type Gender = 'male' | 'female' | 'n/a' | 'none';
export type Color = string;
export type URL = string;

export interface Character {
  name: string;
  height: string;
  mass: string;
  hair_color: Color;
  skin_color: Color;
  eye_color: Color;
  birth_year: string;
  gender: Gender;
  homeworld: URL;
  films: URL[];
  species: URL[];
  vehicles: URL[];
  starships: URL[];
  created: string;
  edited: string;
  url: URL;
}
