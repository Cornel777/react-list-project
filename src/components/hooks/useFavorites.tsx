import {useState} from 'react'
import { Character } from '../../types/Character';


export const STORAGE_KEY = 'SW_favorites';

interface HookReturn {
  favorites: Character[];
  toggleFavorite: (character: Character) => void;
  isFavorite: (character: Character) => boolean;
}

export const useFavorites = (): HookReturn => {
  const [favorites, setFavorites] = useState<Character[]>(() => {
    try {
      const stored = localStorage.getItem(STORAGE_KEY);
      return stored ? JSON.parse(stored) : [];
    } catch (error) {
      console.error('Error encoutered while fetching favorites list' , error);
      return [];
    }
});

  const toggleFavorite = (character: Character) => {
    try {
      setFavorites((prevFavorites) => {
        const isStored = prevFavorites.some((favorite) => favorite.url === character.url);

        let newFavorites: Character[];
        if (isStored) {
          newFavorites = prevFavorites.filter((favorite) => favorite.url !== character.url);
        } else {
          newFavorites =  [...prevFavorites, character];
        }
        
        localStorage.setItem(STORAGE_KEY, JSON.stringify(newFavorites));
        return newFavorites;
      })
    } catch (error) {
      console.error('Error encountered while toggling favorite', error);
    }
  }

  const isFavorite = ( character: Character) => {
    return favorites.some((favorite) => favorite.url === character.url);
  }
  return { favorites, toggleFavorite, isFavorite };
}

export default useFavorites;