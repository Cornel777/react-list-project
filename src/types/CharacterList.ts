import { Character } from "./Character";
import { URL } from "./Character";

export interface CharacterList {
  count: number;
  next: URL | null;
  previous: URL | null;
  results: Character[];
}