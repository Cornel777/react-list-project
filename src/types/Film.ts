import { URL } from './Character';

export interface Film {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  characters: URL[];
  planets: URL[];
  starships: URL[];
  vehicles: URL[];
  species: URL[];
  created: Date;
  edited: Date;
  url: URL;
}