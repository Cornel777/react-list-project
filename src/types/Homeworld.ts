import { URL } from "./Character";

export interface Homeworld {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: URL[];
  created: string;
  edited: string;
  url: URL;
}

