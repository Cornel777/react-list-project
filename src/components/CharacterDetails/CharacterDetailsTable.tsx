import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import Homeworld from "./components/Homeworld/Homeworld";
import { StyledTableRow, StyledTableCell } from "./components/styled/styledTable";
import { FunctionComponent } from "react";
import { Character } from "../../types/Character";

export interface CharacterDetailsTableProps {
  data: Character
}

const CharacterDetailsTable: FunctionComponent<CharacterDetailsTableProps> = ({ data }) => {
  return (
    <TableContainer component={Paper} style={{ maxWidth: '400px' }}>
        <Table size="medium" aria-label="Character traits table">
          <TableHead>
            <StyledTableRow>
              <StyledTableCell>Character trait</StyledTableCell>
              <StyledTableCell>Value</StyledTableCell>
            </StyledTableRow>
          </TableHead>
          <TableBody>
            <StyledTableRow>
              <StyledTableCell>Hair colour</StyledTableCell>
              <StyledTableCell>{data.hair_color}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Eye colour</StyledTableCell>
              <StyledTableCell>{data.eye_color}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Gender</StyledTableCell>
              <StyledTableCell>{data.gender}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Homeworld</StyledTableCell>
              <StyledTableCell>
                <Homeworld homeworldUrl={data.homeworld} />
              </StyledTableCell>
            </StyledTableRow>
          </TableBody>
        </Table>
      </TableContainer>
  )
}

export default CharacterDetailsTable;