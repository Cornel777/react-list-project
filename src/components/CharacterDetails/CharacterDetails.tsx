import {
  Typography,
  Button,
} from '@mui/material'
import FavoriteIcon from '@mui/icons-material/Favorite';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { useParams, useNavigate, useLocation } from 'react-router'
import LoadingDots from '../LoadingDots/LoadingDots'
import { Character } from '../../types/Character';
import BreakElement from './components/styled/BreakElement';
import NameControlButtonsContainer from './components/styled/NameControlButtonsContainer';
import CharacterDetailsWrapperContainer from './components/styled/CharacterDetailsWrapperContainer';
import { FunctionComponent, useEffect, useRef } from 'react';
import useFavorites from '../hooks/useFavorites';
import FilmsTable from './components/Film/FilmsTable';
import StarshipsTable from './components/Starship/StarshipsTable';
import CharacterDetailsTable from './CharacterDetailsTable';
import useFetchQuery from './hooks/useFetchQuery';

const CharacterDetails: FunctionComponent = () => {
  const backButtonRef = useRef<HTMLButtonElement>(null);
  const navigate = useNavigate();
  const location = useLocation();
  const { page } = location.state || { page: 1 };
  const urlParams = useParams();
  const characterId = urlParams.characterId;

  const { isFavorite, toggleFavorite } = useFavorites();
 
  const { isPending, isError, error, data} = useFetchQuery<Character>('characterDetails', `people/${characterId}/`, characterId);

  useEffect(() => {
    backButtonRef.current?.focus();
  }, []);

 
  const handleBackNavigation = () => {
    navigate(`/characters`, { 
      state: { page },
      replace: true
    });
  };
  const onKeyDownHandlerNavigation = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter' || e.key === ' ') {
      handleBackNavigation();
    }
  };

  const handleToggleFavorite = (favorite: Character) => {
    toggleFavorite(favorite);
  }
  const handleOnKeyUptoggleFavorite = (e: React.KeyboardEvent, favorite: Character) => {
    if (e.key === 'Enter' || e.key === ' ') {
      handleToggleFavorite(favorite);
    }
  };

  if (isPending) {
    return <LoadingDots dataTestId="three-loading-dots" />;
  }
  if (isError) {
    return <p>`Failed to load data for character ${error && error.message}`</p>;
  }
  if (!data) {
    return null;
  }

  return (
    <CharacterDetailsWrapperContainer>
      <NameControlButtonsContainer>
        <Button
          aria-label="Go Back"
          data-testId="character-back-button"
          ref={backButtonRef} 
          onClick={handleBackNavigation}
          onKeyDown={(e) => onKeyDownHandlerNavigation(e)}
        >
          <ArrowBackIcon />
        </Button>
        <Typography variant="h3">{data.name}</Typography>
        <Button 
          aria-label="toggle favorites button"
          startIcon={
            <FavoriteIcon 
              color={isFavorite(data) ? 'error' : 'inherit'}
            />
          } 
          onClick={() => handleToggleFavorite(data)}
          onKeyUp={(e) => handleOnKeyUptoggleFavorite(e, data)}
          size='large'
        />
      </NameControlButtonsContainer>
      <CharacterDetailsTable data={data} />
      <BreakElement />
      <Typography variant="h5">Films</Typography>
      <BreakElement/>
      <FilmsTable data={data} />
      <BreakElement />
      <Typography variant="h5">Starships</Typography>
      <BreakElement/>
      <StarshipsTable data={data} />
    </CharacterDetailsWrapperContainer>
  )
}

export default CharacterDetails
