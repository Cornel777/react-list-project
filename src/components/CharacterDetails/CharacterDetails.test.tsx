import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import CharacterDetails from './CharacterDetails';
import useFetchQuery, { HookReturn } from './hooks/useFetchQuery';
import useFavorites from '../hooks/useFavorites';
import { Character } from '../../types/Character';

vi.mock('./hooks/useFetchQuery');
vi.mock('../hooks/useFavorites');

const mockNavigate = vi.fn();
vi.mock('react-router', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => mockNavigate,
    useLocation: vi.fn().mockReturnValue({ state: { page: 1 } }),
    useParams: vi.fn().mockReturnValue({ characterId: '1' })
  };
});

describe('CharacterDetails', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  beforeEach(() => {
    queryClient.clear();
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: {
        name: 'Luke Skywalker',
        url: 'https://swapi.dev/api/people/1/',
        height: '172',
        gender: 'male',
        homeworld: 'url1',
        films: [],
        species: [],
        vehicles: [],
        starships: [],
        created: '',
        edited: '',
        birth_year: '',
        eye_color: '',
        hair_color: '',
        skin_color: '',
        mass: ''
      }
    });

    vi.mocked(useFavorites).mockReturnValue({
      isFavorite: vi.fn().mockReturnValue(false),
      toggleFavorite: vi.fn(),
      favorites: []
    });
    mockNavigate.mockClear();
  });

  

  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }

  it('matches snapshot', () => {
    const { container } = renderWithClient(<CharacterDetails />);
    expect(container).toMatchSnapshot();
  });

  it('shows loading state', () => {
    vi.mocked(useFetchQuery).mockReturnValueOnce({
      isPending: true,
      isError: false,
      error: null,
      data: undefined
    } as HookReturn<Character>);

    renderWithClient(<CharacterDetails />);
    expect(screen.getByTestId('three-loading-dots')).toBeInTheDocument();
  });

  it('shows error state', async () => {
    vi.mocked(useFetchQuery).mockReturnValueOnce({
      isPending: false,
      isError: true,
      error: new Error('Error loading character'),
      data: undefined,
      isFetching: false,
      isPlaceholderData: false
    });

    renderWithClient(<CharacterDetails />);
    expect(await screen.findByText(/Failed to load data for character/i)).toBeInTheDocument();
  });

  it('renders heading names', async () => {
    const result = renderWithClient(<CharacterDetails />);
    expect((await result.getByRole('heading', {name: /Luke Skywalker/i }))).toBeInTheDocument();
    expect((await result.getByRole('heading', {name: /Films/i }))).toBeInTheDocument();
    expect((await result.getByRole('heading', {name: /Starships/i }))).toBeInTheDocument();
  });

  it('navigates back correctly', async () => {
    renderWithClient(<CharacterDetails />);
    const backButton = await screen.findByTestId('character-back-button');
    fireEvent.click(backButton);

    expect(mockNavigate).toHaveBeenCalledWith('/characters', {
      replace: true,
      state: { page: 1 },
    });
  });

  it('toggles favorite on button click', async () => {
    const mockToggle = vi.fn();
    vi.mocked(useFavorites).mockReturnValueOnce({
      isFavorite: vi.fn(() => false),
      toggleFavorite: mockToggle,
      favorites: []
    });

    renderWithClient(<CharacterDetails />);
    const favButton = await screen.findByRole('button', { name: /favorite/i });
    fireEvent.click(favButton);

    expect(mockToggle).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'Luke Skywalker'
      })
    );
  });
});