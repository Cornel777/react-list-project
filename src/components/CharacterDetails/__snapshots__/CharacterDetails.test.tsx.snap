// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterDetails > matches snapshot 1`] = `
<div>
  <div
    class="sc-gaZyOd ghTcZX"
  >
    <div
      class="sc-eaUbBy izwuGO"
    >
      <button
        aria-label="Go Back"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
        data-testid="character-back-button"
        id=":r0:"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
          />
        </svg>
      </button>
      <h3
        class="MuiTypography-root MuiTypography-h3 css-13msox4-MuiTypography-root"
      >
        Luke Skywalker
      </h3>
      <button
        aria-label="toggle favorites button"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorPrimary css-1cg7qmj-MuiButtonBase-root-MuiButton-root"
        id=":r1:"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 400px;"
    >
      <table
        aria-label="Character traits table"
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-17e8nb1-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Character trait
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Hair colour
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            />
          </tr>
          <tr
            class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Eye colour
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            />
          </tr>
          <tr
            class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Gender
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              male
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Homeworld
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Luke Skywalker
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <br
      class="sc-gyycJP"
    />
    <h5
      class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
    >
      Films
    </h5>
    <br
      class="sc-gyycJP"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 400px;"
    >
      <table
        aria-label="Films table"
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-17e8nb1-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Film name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Episode no.
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        />
      </table>
    </div>
    <br
      class="sc-gyycJP"
    />
    <h5
      class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
    >
      Starships
    </h5>
    <br
      class="sc-gyycJP"
    />
    <p>
      No starships flown
    </p>
  </div>
</div>
`;
